### ISSUE TITLE
Markdown syntax highlighting bug due to a dollar sign ($) within a code block

### ISSUE THREAD

**Issue by SultanOrazbayev** [jupyterlab/jupyterlab#6774]
**Describe the bug**

The syntax highlighting breaks down once a dollar sign is used within a code block (see the attached image). I tried using syntax highlighting (e.g. specifying `bash` after the triple backticks), but it doesn't help.

[Screen Shot 2019-07-06 at 17.20.40.pdf](https://github.com/jupyterlab/jupyterlab/files/3365050/Screen.Shot.2019-07-06.at.17.20.40.pdf)

**To Reproduce**

1. Create a new Markdown file and open it with the editor in Jupyter lab.
2. Type the following code:

```
# Hello

Sample code:

    ```
    echo $HOME
    ```

# Next section

Note: no syntax highlighting above.

```

3. See error (the section title is not highlighted after the dollar sign).

**Expected behavior**
The syntax highlighting of the `# Next section` and any supporting code below.

**Desktop (please complete the following information):**

- OS: MacOS
- Browser: Chrome 75
- JupyterLab 1.0.1

** Potentially related cases (jupyter notebook)

https://github.com/jupyterlab/jupyterlab/issues/1834

#5266
#547

**Comment by SultanOrazbayev** [comment]
To add: the markdown renders and displays correctly, it's only syntax highlighting within the jupyterlab editor.

**Comment by ian-r-rose** [comment]
Thanks for the report @SultanOrazbayev! I have confirmed that the CodeMirror demo [here](https://codemirror.net/mode/markdown/) does not display this rendering problem, so the bug is on our end (n.b., the classic notebook also shows this bug).

This is probably a mistake in our multiplexing codemirror mode here: https://github.com/jupyterlab/jupyterlab/blob/079115dd36705f11067ebfd9c34f783719f1775a/packages/codemirror/src/codemirror-ipythongfm.ts#L32-L58
which switches to LaTeX mode upon encountering a `$` symbol.

I'm not sure what the fix would look like just now -- we would need to identify that we are in a code fence and not turn on LaTeX mode.

**Comment by SultanOrazbayev** [comment]
Thanks for the follow-up!

I'm not sure if this is an option, but could a keyword at the beginning of the code block be used to disable texMode? e.g. if the code block begins with ```shell  or ```console or ```bash?

**Comment by austincauthon** [comment]
A workaround for this issue is to use the unicode `&#0036;` in place of using a `$`. This will render a `$` inside your rendered document, but not affect your syntax highlighting in your editor.

**Comment by RA80533** [comment]
I encountered this strange quirk today. Phew, I'm glad we know where it's coming from at least.

**Comment by fperez** [comment]
Confirming that this remains in current JLab (3.0.12) but doesn't occur in classic, same file open side-by-side simultaneously in both:

![image](https://user-images.githubusercontent.com/57394/113363851-dbb39b80-9306-11eb-9612-238c668dbaca.png)

**Comment by telamonian** [comment]
I wonder if this is related to #8645. Both may boil down to issues with detecting where you're trying to enter into "mathmode"

**Comment by fperez** [comment]
@telamonian agreed - I haven't looked at the logic, but basically the math mode detection shouldn't cross code  boundaries, nicely highlighted by having two:

![image](https://user-images.githubusercontent.com/57394/113366981-abbcc600-930f-11eb-8c2b-cf6c9b7064a3.png)

**Comment by tannerlegvold** [comment]
Just as a further note, this also occurs in JupyterLab 3.0.14 using `<code> </code> ` blocks instead of backtick code blocks.

**Comment by telamonian** [comment]
For reference, the way the mathmode parsing works is that we take this Very Special regex:

https://github.com/jupyterlab/jupyterlab/blob/32c98cc091537566e79d98a40c9950381e38978a/packages/rendermime/src/latex.ts#L16

and use it to split markdown cells into non-math and math parts. It's likely that in order to fix this issue either `MATHSPLIT` or the code that directly consumes it will have to be fixed.

**Comment by RA80533** [comment]
Hm, that's interesting because, as @ian-r-rose noted, there's functionality for changing the context to TeX upon encountering the delimiter elsewhere (see [codemirror-ipythongfm.ts](https://github.com/jupyterlab/jupyterlab/blob/850445a3685aecac4c8a0dd942709ad212e248fc/packages/codemirror/src/codemirror-ipythongfm.ts)). I suspect the regex is unnecessary for the reason of CodeMirror having built-in support for this kind of behavior (which, as aforementioned, appears to have been implemented).

**Comment by mattlegro** [comment]
I don't know if this is useful information, but even using the unicode workaround doesn't fix the rendering issue. Text between two `&#0036;` also is italicised and bolded. In the following example, it renders "15 billion.....at" as bold and italicized,

```
* Netflix is reported to invest &#0036;15 billion on new content; Prime Video and Apple TV+ at &#0036;6 billion; Hulu, Disney+ and HBO Max are all coming in around &#0036;3 billion
```

**Comment by krassowski** [comment]
I looked into this one. The clean solution would be to re-write the `CodeMirror.multiplexingMode` to use the context of the outer mode (`gfmMode`) when deciding whether to proceed with switching to the inner mode, but this is a difficult thing as we would need to manage concurrent parsers (modes) consuming the same stream (as does the `overlayMode`) which might end up very error-prone (and might not be accepted upstream given the slow migration to CodeMirror 6). I found a simple workaround which should make everyone happy for now, hopefully without adding maintenance overhead - see #10867.

**Comment by krassowski** [comment]
There are further fixes to the remaining edge cases in https://github.com/jupyterlab/jupyterlab/pull/11479 and https://github.com/jupyterlab/jupyterlab/pull/11488 - hopefully both will get included in the next 3.2.x release. Please help testing (e.g. on binder) if you can. If you discover new edge cases after release please open a new issue and tag me.
